subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/JdkTools.java

```diff
@@ -103,7 +108,18 @@ private JavaCompiler buildJavaCompiler() {
             if (isJava9Compatible) {
                 clazz = isolatedToolsLoader.loadClass("javax.tools.ToolProvider");
                 try {
-                    JavaCompiler compiler = (JavaCompiler) clazz.getDeclaredMethod("getSystemJavaCompiler").invoke(null);
+                    JavaCompiler compiler = null;
+                    for (JavaCompiler compilerCandidate : ServiceLoader.load(JavaCompiler.class, isolatedToolsLoader)) {
+                        if (!compilerCandidate.getClass().getName().equals(DEFAULT_COMPILER_IMPL_NAME)) {
+                            // take the first compiler found on the custom compiler classpath that is not the JavacTool system compiler (if any)
+                            compiler = compilerCandidate;
```

Comment:
I think this should probably at least log which compiler is selected to debug, to enable troubleshooting if multiple compilers are accidentally included.